core/src/main/scala/org/apache/spark/SparkContext.scala

```diff
@@ -2723,44 +2689,50 @@ object SparkContext extends Logging {
+            " needs to be specified since a task requirement config: " +
+            ResourceID(SPARK_TASK_PREFIX, taskReq.resourceName).amountConf +
+            " was specified")
         )
         // Make sure the executor resources are large enough to launch at least one task.
-        if (execCount.toLong < taskCount.toLong) {
-          throw new SparkException(
-            s"The executor resource config: " +
-              s"${SPARK_EXECUTOR_RESOURCE_PREFIX + rName + SPARK_RESOURCE_AMOUNT_SUFFIX} " +
-              s"= $execCount has to be >= the task config: " +
-              s"${SPARK_TASK_RESOURCE_PREFIX + rName + SPARK_RESOURCE_AMOUNT_SUFFIX} = $taskCount")
+        if (execCount < taskReq.amount) {
+          throw new SparkException("The executor resource config: " +
+            ResourceID(SPARK_EXECUTOR_PREFIX, taskReq.resourceName).amountConf +
+            s" = $execCount has to be >= the task config: " +
+            ResourceID(SPARK_TASK_PREFIX, taskReq.resourceName).amountConf +
+            s" = ${taskReq.amount}")
         }
         // Compare and update the max slots each executor can provide.
-        val resourceNumSlots = execCount.toInt / taskCount
+        val resourceNumSlots = execCount / taskReq.amount
         if (resourceNumSlots < numSlots) {
+          logWarning(s"The configuration of resource: ${taskReq.resourceName} " +
+            s"(limits tasks to $resourceNumSlots) will result in wasted resources of resource " +
+            s"${limitingResourceName} (would allow for $numSlots tasks). " +
+            "Please adjust your configuration.")
           numSlots = resourceNumSlots
-          limitingResourceName = rName
+          limitingResourceName = taskReq.resourceName
         }
       }
       // There have been checks above to make sure the executor resources were specified and are
       // large enough if any task resources were specified.
-      taskResourcesAndCount.foreach { case (rName, taskCount) =>
-        val execCount = executorResourcesAndCounts(rName)
-        if (taskCount.toInt * numSlots < execCount.toInt) {
-          val message = s"The configuration of resource: $rName (exec = ${execCount.toInt}, " +
-            s"task = ${taskCount}) will result in wasted resources due to resource " +
-            s"${limitingResourceName} limiting the number of runnable tasks per executor to: " +
-            s"${numSlots}. Please adjust your configuration."
+      taskResourceRequirements.foreach { taskReq =>
+        val execCount = executorResourcesAndCounts(taskReq.resourceName)
+        if (taskReq.amount * numSlots < execCount) {
+          val message = s"The configuration of resource: ${taskReq.resourceName} " +
+            s"(exec = ${execCount}, task = ${taskReq.amount}) will result in wasted " +
+            s"resources due to resource ${limitingResourceName} limiting the number of " +
+            s"runnable tasks per executor to: ${numSlots}. Please adjust your configuration."
```

Comment:
I'd prefer to log warn the limited `numSlots` right above `taskResourceRequirements.foreach` if `limitingResourceName` changed. And in `message`, appending the specific amount of the wasted resource to show user the config result directly.

Reply:
I think it is unnecessary because in most use cases we will only see one CPU and one additional resource.